lib/ex_unit/lib/ex_unit/capture_server.ex

```diff
@@ -83,25 +84,71 @@ defmodule ExUnit.CaptureServer do
     {:noreply, config}
   end
 
-  defp release_device(ref, %{devices: {names, refs}} = config) do
-    case Map.pop(refs, ref) do
-      {{name, pid}, refs} ->
-        names = Map.delete(names, name)
-        Process.demonitor(ref, [:flush])
-
-        try do
-          try do
-            Process.unregister(name)
-          after
-            Process.register(pid, name)
-          end
-        rescue
-          ArgumentError -> nil
-        end
+  defp capture_device(name, encoding, "", %{devices: devices} = config, caller)
+       when is_map_key(devices, name) do
+    case Map.get(devices, name) do
+      %{encoding: ^encoding} = device ->
+        {_, output} = StringIO.contents(device.pid)
+        ref = Process.monitor(caller)
+        refs = [{ref, byte_size(output)} | device.refs]
+        devices = %{devices | name => %{device | refs: refs}}
+        {:reply, {:ok, ref}, %{config | devices: devices}}
+
+      %{encoding: other_encoding} ->
+        {:reply, {:error, {:changed_encoding, other_encoding}}, config}
+    end
+  end
+
+  defp capture_device(name, _, _, %{devices: devices} = config, _)
+       when is_map_key(devices, name) do
+    {:reply, {:error, :input_on_already_captured_device}, config}
+  end
+
+  defp capture_device(name, encoding, input, %{devices: devices} = config, caller) do
+    {:ok, pid} = StringIO.open(input, encoding: encoding)
+    orig_pid = Process.whereis(name)
 
-        %{config | devices: {names, refs}}
+    Process.unregister(name)
+    Process.register(pid, name)
 
-      {nil, _refs} ->
+    _ = Process.monitor(pid)
+    ref = Process.monitor(caller)
+
+    device = %{orig_pid: orig_pid, pid: pid, refs: [{ref, 0}], encoding: encoding}
+    devices = Map.put(devices, name, device)
+    {:reply, {:ok, ref}, %{config | devices: devices}}
+  rescue
+    ArgumentError ->
+      {:reply, {:error, :no_device}, config}
+  end
+
+  defp release_device(ref, %{devices: devices} = config) do
+    case device_by_ref(devices, ref) do
+      {name, device_info} ->
+        case Enum.reject(device_info.refs, &refs_equal?(&1, ref)) do
+          [] ->
+            try do
+              try do
+                Process.unregister(name)
+              after
+                Process.register(device_info.orig_pid, name)
+                StringIO.close(device_info.pid)
+                Process.demonitor(ref)
```

Comment:
It is best to call this function only on `device_capture_off`, since in the case of the DOWN message it will already be demonitored. You also want to pass the flush argument:

    Process.demonitor(ref, [:flush])

So the :DOWN message is discarded.